samples/mpu/mem_domain_apis_test/src/main.c

```diff
@@ -0,0 +1,128 @@
+/* main.c - Memory Domain APIs demo */
+
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <misc/printk.h>
+
+/* size of stack area used by each thread */
+#define STACKSIZE 1024
+
+/* scheduling priority used by each thread */
+#define PRIORITY 7
+
+/* busy waiting time (in us) */
+#define BUSY_WAITING_TIME 3000000
+
+
+K_THREAD_STACK_ARRAY_DEFINE(app_stack, 3, STACKSIZE);
+
+struct k_thread app_thread_id[3];
+
+struct k_mem_domain app_domain[2];
+
+/* the start address of the MPU region needs to align with its size */
+u8_t __aligned(32) app0_buf[32];
+u8_t __aligned(32) app1_buf[32];
+
+struct k_mem_partition app0_parts[] = {
+	MEM_PARTITION_ENTRY((u32_t)app0_buf,
+			    sizeof(app0_buf),
+			    K_MEM_PARTITION_P_RW_U_RW |
+			    K_MEM_PARTITION_EXECUTE_NEVER),
+	MEM_PARTITION_ENTRY((u32_t)app1_buf,
+			    sizeof(app1_buf),
+			    K_MEM_PARTITION_P_RW_U_RO |
+			    K_MEM_PARTITION_EXECUTE_NEVER),
+};
+
+struct k_mem_partition app1_parts[] = {
+	MEM_PARTITION_ENTRY((u32_t)app1_buf,
+			    sizeof(app1_buf),
+			    K_MEM_PARTITION_P_RW_U_RW |
+			    K_MEM_PARTITION_EXECUTE_NEVER),
+	MEM_PARTITION_ENTRY((u32_t)app0_buf,
+			    sizeof(app0_buf),
+			    K_MEM_PARTITION_P_RW_U_RO |
+			    K_MEM_PARTITION_EXECUTE_NEVER),
+};
+
+/* application thread */
+void app_thread(void *dummy1, void *dummy2, void *dummy3)
+{
+	int id = (int) dummy1;
+	int count = 0;
+
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	printk("app thread %d (%p) starts.\n", id, &app_thread_id[id]);
+
+	/*
+	 * app thread 0 will be added into app0_domain.
+	 * app thread 1 will be added into app1_domain.
+	 * app thread 2 will be added into app2_domain.
+	 */
+	printk("app thread %d (%p) adds itself into app%d_domain.\n",
+		id, &app_thread_id[id], id%2);
+
+	k_mem_domain_add_thread(&app_domain[id%2],
+				&app_thread_id[id]);
+
+	k_busy_wait(BUSY_WAITING_TIME);
+
+	while (1) {
+		printk("app thread %d (%p) yields.\n", id, &app_thread_id[id]);
+
+		k_yield();
+
+		printk("app thread %d (%p) is back.\n", id, &app_thread_id[id]);
+
+		k_busy_wait(BUSY_WAITING_TIME);
+
+		if (count == 4 && id == 1) {
```

Comment:
> To get the best demonstration for memory domain APIs, we need to enable unprivileged threads
> execution environment first.

That's only true when using the ARM MPU because the Zephyr configuration for it sets the PRIVDEFENA bit. It's not necessary to set this bit, though, and if it is clear, the MPU will restrict access for even privileged threads for memory areas that are not otherwise defined in the regions file for the SoC. That is, clearing PRIVDEFENA on ARM targets with the ARM MPU would also let you show some example uses of this API along with the RFC without enabling unprivileged threads, which would make it easier to understand.

Honestly I think it is a mistake in general to set PRIVDEFENA: even privileged code should not be exempt from the MPU regions specifying exactly what areas we want accessed, and how.

I tested disabling this bit with this patch: https://hastebin.com/edimagivup.cpp

Unfortunately, some early SoC init code on nRF chips failed because it accesses memory outside of the MPU memory regions, and I've been waiting for this series to get posted so we can start cleaning this up :).

If you have any boards (96b_carbon, 96b_nitrogen, for example) which have the ARM MPU, you can apply the above patch to disable PRIVDEFENA. Then your series can try to access memory (like the system memory I mentioned earlier on 96b_carbon) with and without a memory domain defined which gives it access. Even a privileged thread should crash without the right memory domain configuration.